scheduler-worker/src/main/java/gov/nasa/jpl/aerie/scheduler/worker/services/SynchronousSchedulerAgent.java

```diff
@@ -254,6 +258,31 @@ public void schedule(final ScheduleRequest request, final ResultsProtocol.Writer
     }
   }
 
+  private Optional<DatasetId> storeSimulationResults(PlanningHorizon planningHorizon, SimulationFacade simulationFacade, PlanMetadata planMetadata,
+                                                     final Map<SchedulingActivityDirective, ActivityDirectiveId> schedulingActivityDirectiveToMerlinActivityDirectiveIds)
+  throws PlanServiceException, IOException
+  {
+    //finish simulation until end of horizon before posting results
+    simulationFacade.computeSimulationResultsUntil(planningHorizon.getEndAerie());
+    final var schedulingActivityDirectiveToMerlinActivityDirective = schedulingActivityDirectiveToMerlinActivityDirectiveIds
+        .entrySet()
+        .stream()
+        .collect(Collectors.toMap((a) -> new SchedulingActivityDirectiveId(a.getKey().id().id()), Map.Entry::getValue));
+    final var simulationActivityCorrespondance = simulationFacade.getActivityIdCorrespondance();
+    final var simulationActivityDirectiveIdToMerlinActivityDirectiveId =
+        simulationActivityCorrespondance.entrySet().stream().collect(Collectors.toMap(
+        Map.Entry::getValue,
+        (a) -> schedulingActivityDirectiveToMerlinActivityDirective.get(a.getKey())));
+    if(simulationActivityDirectiveIdToMerlinActivityDirectiveId.values().equals(schedulingActivityDirectiveToMerlinActivityDirectiveIds.values())){
```

Comment:
The names, while verbose, do not actually match the types of the objects. Specifically:

- `schedulingActivityDirectiveToMerlinActivityDirective` is a `Map<SchedulingActivityDirectiveId, ActivityDirectiveId>`
- `simulationActivityDirectiveIdToMerlinActivityDirectiveId` is a `Map<ActivityDirectiveId, ActivityDirectiveId>`

As such, I'm not confident that this method is correct. Could you go over exactly how this check is supposed to work?

Reply:
the first map has all the merlin ids of currently simulated activities in its values. The second map is from scheduling plan to merlin ids.  So if both sets are equal, we know all the activities in the plan are simulated, don't we ?

Reply:
1) How does this interact with decomposing activities? 
2) The variable names are confusing. It wasn't until I renamed everything that I was able to follow how all of the remappings worked.

```java
private Optional<DatasetId> storeSimulationResults(PlanningHorizon planningHorizon, SimulationFacade simulationFacade, PlanMetadata planMetadata,
                                                     final Map<SchedulingActivityDirective, ActivityDirectiveId> schedDirectiveToMerlinId)
  throws PlanServiceException, IOException
  {
    //finish simulation until end of horizon before posting results
    simulationFacade.computeSimulationResultsUntil(planningHorizon.getEndAerie());
    final var schedID_to_MerlinID =
        schedDirectiveToMerlinId.entrySet().stream()
        .collect(Collectors.toMap(
            (a) -> new SchedulingActivityDirectiveId(a.getKey().id().id()), Map.Entry::getValue));
    final var schedID_to_simID =
        simulationFacade.getActivityIdCorrespondance();
    final var simID_to_MerlinID =
        schedID_to_simID.entrySet().stream().collect(Collectors.toMap(
        Map.Entry::getValue,
        (a) -> schedID_to_MerlinID.get(a.getKey())));
    if(simID_to_MerlinID.values().equals(schedDirectiveToMerlinId.values())){
      return Optional.of(planService.storeSimulationResults(planMetadata,
                                                simulationFacade.getLatestDriverSimulationResults(),
                                                simID_to_MerlinID));
    } else{
      //schedule in simulation is inconsistent with current state of the plan (user probably disabled simulation for some of the goals)
      return Optional.empty();
    }
  }
```

Reply:
`schedID_to_simID` only contains ids of activity directives. Child activities ids are not stored. So we check only the correspondence for activity directives. 
Then for storing results, we proceed exactly like the merlin server. Thus, spans of child activities are stored.